Add support for "More Actions" column in Resource Viewer

Charles-Gagnon opened the pull request:
This allows adding a "More Actions" column which will be a button that when clicked opens a context menu with the list of commands to run.

Currently only supporting running commands - may add support for more things in the future as needed.

Customization of the column is purposely limited - most of the settings are hard-coded since they don't make sense for this column (such as sorting or filtering)

I also cleaned up some of the typings to refer to the correct types actually being used by the grid. 

![image](https://user-images.githubusercontent.com/28519865/97349835-32dfac80-184d-11eb-899a-f8fcf8d6fe1b.png)

alanrenmsft reviewed (changes requested):
I think we should leverage the menu contribution point for this featur:, define a new menu id and let extensions contribute menu items. 
aside from clicking on the button, the actions should also be available as context menu for the data grid row, that would be helpful.

alanrenmsft commented on extensions/azurecore/src/azureDataGridProvider.ts:
```diff
@@ -72,7 +75,8 @@ export class AzureDataGridProvider implements azdata.DataGridProvider {
 			{ id: 'location', type: 'text', field: 'locationDisplayName', name: loc.location, width: 150 },
-			{ id: 'subscriptionId', type: 'text', field: 'subscriptionName', name: loc.subscription, width: 150 }
+			{ id: 'subscriptionId', type: 'text', field: 'subscriptionName', name: loc.subscription, width: 150 },
+			{ id: 'actions', type: 'actions', field: '', name: '' }
```
maybe we don't need to ask user to define an actions column, there can be a table wide setting, e.g. showActionsColumn, and then you can get the contributed actions when the button is pressed.

Charles-Gagnon replied:
Yeah that's what I was thinking in my comment above. I was trying to avoid having a bunch of extra options like that but I think it's reasonable to have provider-wide options like that.

Charles-Gagnon replied:
I ended up just making the actions column always show. For the work I'm doing that'll always be the case - and then later on when/if we decide we need to support it not being shown we can do the work then to allow that.

alanrenmsft commented on src/sql/workbench/browser/editor/resourceViewer/resourceViewerInput.ts:
```diff
@@ -47,7 +52,15 @@ export class ResourceViewerInput extends EditorInput {
 			sortable: false
 		});
 		this._register(this._actionsColumn.onClick(args => {
```
not interested in showing this also when context menu is requested?

Charles-Gagnon replied:
Added 😄

Charles-Gagnon commented on src/sql/azdata.proposed.d.ts:
```diff
@@ -226,14 +234,11 @@ declare module 'azdata' {
+		 * The other properties that will be displayed in the grid columns
 		 */
-		[key: string]: string | DataGridHyperlinkInfo;
+		fieldValues: { [key: string]: string | DataGridHyperlinkInfo }
```
Sorry about other things getting here - was making some other changes as I went along and it'd be too much of a pain to try and separate them out. It's usually pretty simple changes though.

Charles-Gagnon commented:
Ok @alanrenmsft should be done with my changes now - added support for context menu as well otherwise things should be mostly the same (I moved some code around so the Input isn't doing UI stuff)

ranasaria commented:
> This allows adding a "More Actions" column which will be a button that when clicked opens a context menu with the list of commands to run.

Is it possible to have a 'context menu' show up on right click on a subscription value?

ranasaria commented on extensions/azurecore/src/extension.ts:
```diff
@@ -91,6 +91,9 @@ export async function activate(context: vscode.ExtensionContext): Promise<azurec
 	registerAzureResourceCommands(appContext, [azureResourceTree, connectionDialogTree]);
 	azdata.dataprotocol.registerDataGridProvider(new AzureDataGridProvider(appContext));
+	vscode.commands.registerCommand('azure.dataGrid.openInAzurePortal', async (item: azdata.DataGridItem) => {
+		await vscode.env.openExternal(vscode.Uri.parse('https://microsoft.com'));
```
Is this is temporary implementation? If so then suggest adding a TODO comment

Charles-Gagnon replied:
Yes it's temporary. I personally don't find TODOs useful and will be replacing this as my next task so didn't bother putting anything there.

ranasaria commented on src/sql/workbench/browser/editor/resourceViewer/resourceViewerInput.ts:
```diff
@@ -3,34 +3,44 @@
 		super();
+		this.actionsColumn = new ButtonColumn<azdata.DataGridItem>({
+			id: 'actions',
+			iconCssClass: 'toggle-more',
```
for my knowledge what does this iconCssClass do? Perhaps a comment indicationg its purpose would not be a bad thing to add.

ranasaria commented on src/sql/workbench/contrib/resourceViewer/browser/resourceViewerTable.ts:
```diff
@@ -107,18 +127,17 @@ export class ResourceViewerTable extends Disposable {
 		if (column) {
 			const row = this._dataView.getItem(event.cell.row);
-			const value = row[column.field];
+			const value = row.fieldValues[column.field];
```
The purpose or significance of this change is not clear to me.

ranasaria reviewed (approved):
Looks good.

alanrenmsft commented on src/sql/workbench/contrib/resourceViewer/browser/resourceViewerEditor.ts:
```diff
@@ -43,6 +54,10 @@ export class ResourceViewerEditor extends EditorPane {
 		const tableContainer = this.createResourceViewerTable();
 
+		this._register(this._resourceViewerTable.onContextMenu(e => {
+			this.showContextMenu(e.anchor, e.item);
```
Could you please check whether this event will be fired for header row as well? if so, need to make sure e.item is not undefined before calling the showContextMenu method.

Charles-Gagnon replied:
Nope - not called for header clicks.
